core/src/main/java/com/datastax/oss/driver/internal/core/data/DefaultTupleValue.java

```diff
@@ -93,6 +95,44 @@ private void readObject(ObjectInputStream stream) throws InvalidObjectException
     throw new InvalidObjectException("Proxy required");
   }
 
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+
+    if (o == null || !(o instanceof TupleValue)) return false;
+    DefaultTupleValue that = (DefaultTupleValue) o;
+
+    if (this.protocolVersion() != that.protocolVersion()) return false;
+
+    for (int i = 0; i < values.length; i++) {
+      DataType innerThisType = type.getComponentTypes().get(i);
+      DataType innerThatType = type.getComponentTypes().get(i);
+      if (!innerThisType.equals(innerThatType)) {
+        return false;
+      }
+      Object thisValue =
+          this.codecRegistry()
+              .codecFor(innerThatType)
+              .decode(this.values[i], this.protocolVersion());
+      Object thatValue =
+          that.codecRegistry()
+              .codecFor(innerThatType)
+              .decode(that.values[i], that.protocolVersion());
+      if (!((thisValue == thatValue) || (thisValue != null && thisValue.equals(thatValue)))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public int hashCode() {
+
+    int result = Objects.hash(type);
+    result = 31 * result + Arrays.hashCode(values);
+    return result;
```

Comment:
If we decode in `equals()`, we must also decode here, otherwise it violates the rule that equal objects must have the same hash codes. For example two equal `varint` values that have different binary representations will be considered equal, but produce different hash codes.